Model/Api/OAuthRedirect.php

```diff
@@ -35,37 +46,198 @@ class OAuthRedirect implements OAuthRedirectInterface
+        $boltPublicKey = $oauthConfiguration['boltPublicKey'];
+
+        $token = $this->ssoHelper->exchangeToken($code, $scope, $clientID, $clientSecret);
+        if ($token === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $payload = $this->ssoHelper->parseAndValidateJWT($token, $clientID, $boltPublicKey);
+        if ($payload === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $websiteId = $this->storeManager->getStore()->getWebsiteId();
+        $storeId = $this->storeManager->getStore()->getId();
+
+        $externalCustomerEntity = null;
+        $customer = null;
+        try {
+            $externalCustomerEntity = $this->externalCustomerEntityRepository->getByExternalID($payload['sub']);
+        } catch (NoSuchEntityException $nsee) {
+        }
+        try {
+            if ($externalCustomerEntity !== null) {
+                $customer = $this->customerRepository->getById($externalCustomerEntity->getCustomerID());
+            } else {
+                $customer = $this->customerRepository->get($payload['email'], $websiteId);
+            }
+        } catch (NoSuchEntityException $nsee) {
+        }
+
+        // External customer entity exists, but the customer it links to doesn't
+        // This should happen very rarely, if at all, which is why we notify bugsnag when this happens
+        if ($externalCustomerEntity !== null && $customer === null) {
+            $this->bugsnag->notifyError(
+                'OAuthRedirect',
+                'external customer entity ' . $payload['sub'] . ' linked to nonexistent customer ' . $externalCustomerEntity->getCustomerID()
+            );
+        }
+
+        // If the customer isn't linked and it exists in M2, but the email is not verified, we throw an exception
+        if ($externalCustomerEntity === null && $customer !== null && !$payload['email_verified']) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        try {
+            $customer = $customer ?: $this->createNewCustomer($websiteId, $storeId, $payload);
+            $this->linkAndLogin($payload['sub'], $customer->getId());
+        } catch (Exception $e) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+    }
+
+    /**
+     * Create new customer from provided parameters
+     *
+     * @param int   $websiteId
+     * @param int   $storeId
+     * @param mixed $payload
+     */
+    private function createNewCustomer($websiteId, $storeId, $payload)
+    {
+        $newCustomer = $this->customerInterfaceFactory->create();
+        $newCustomer->setWebsiteId($websiteId);
+        $newCustomer->setStoreId($storeId);
+        $newCustomer->setFirstname($payload['first_name']);
```

Comment:
Have you checked that first_name exists? Otherwise we have notice there

Reply:
it will definitely exist. see NewSession in oauth/internal/authorize_processor.go

Reply:
I don't think we should rely on go code. It can be changed down the road. If we know it should exists, could we check it and throw an exception if doesn't? You already have this check for other payload elements.

Reply:
that's fair i will make a change